sys/aarch64/evec.S

```diff
@@ -134,6 +162,35 @@ ENTRY(handle_user_trap)
         mov     x0, sp
         bl      user_trap_handler
 user_exc_leave:
+        /* load thread_t::tdp_flags */
+        /* thread_t::tdp_flags is a volatile unsigned - use 32-bit */
+        mrs     x1, tpidr_el1
+        ldr     x1, [x1, #PCPU_CURTHREAD]
+        ldr     w3, [x1, #TD_PFLAGS]
+
+        and     w2, w3, #TDP_FPUINUSE
+        cmp     w2, wzr
+        beq     .skip_fpu_restore
+
+        /* enable FPU */
+        mrs     x2, cpacr_el1
+        and     x2, x2, ~CPACR_FPEN_MASK
+        orr     x2, x2, CPACR_FPEN_TRAP_NONE
+        msr     cpacr_el1, x2
+
+        and     w2, w3, #TDP_FPUCTXSAVED
+        cmp     w2, wzr
+        beq     .skip_fpu_restore
+
+        /* clear TDP_FPUCTXSAVED flag */
+        and     w2, w3, ~TDP_FPUCTXSAVED
+        str     w2, [x1, #TD_PFLAGS]
+
+        /* restore FPU context */
+        ldr     x1, [x1, #TD_UCTX]
+        load_fpu_ctx x1, x2, w2 /* 2nd and 3rd argument must be the same */
```

Comment:
Here I would write sth like '2nd and 3rd argument must be register and its lower half'.